File: rs/src/configuration.rs
/*
 WebServe (Rust port)
 Copyright 2021-2024 Peter Pearson.

 Licensed under the Apache License, Version 2.0 (the "License");
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ---------
*/

#![allow(dead_code)]

use std::fmt;

use std::io::{BufRead, BufReader};

use std::collections::BTreeMap;

#[derive(Clone, Debug)]
pub struct SiteConfig {
pub name:           String,
pub ctype:          String,

pub definition:     String,

pub params:         BTreeMap<String, String>
}

impl Default for SiteConfig {
    fn default () -> SiteConfig {
        SiteConfig{name: String::new(), ctype: String::new(), definition: String::new(),
                    params: BTreeMap::new()}
    }
}

impl fmt::Display for SiteConfig {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Name: {}, type: {}", self.name, self.ctype)
    }
}

impl SiteConfig {

    pub fn has_param(&self, param: &str) -> bool {
        self.params.contains_key(param)
    }

    pub fn get_param(&self, param: &str) -> String {
        let mut ret_value = String::new();

        if self.params.contains_key(param) {
            ret_value = self.params.get(param).unwrap().clone();
        }

        ret_value
    }

    pub fn get_param_as_bool(&self, param: &str, default: bool) -> bool {
        let result = self.params.get(param);
        match result {
            None => return default,
            Some(val) => {
                if val == "1" || val == "true" || val == "yes" {
                    true
                }
                else {
                    false
                }
            }
        }
    }

    pub fn get_param_as_uint(&self, param: &str, default: u32) -> u32 {
        let result = self.params.get(param);
        match result {
            None => default,
            Some(val) => {
                let uint_value = val.parse::<u32>().unwrap();
                uint_value
            }
        }
    }
}

#[derive(Clone, Debug)]
pub struct Configuration {
pub worker_threads:                 u32,
pub enable_http:                    bool,
pub port_number_http:               u32,

pub downgrade_user_after_bind:      bool,
pub downgrade_username:             String,

pub keep_alive_enabled:             bool,
pub keep_alive_timeout:             u32,
pub keep_alive_limit:               u32,

pub site_configs:                   Vec<SiteConfig>
}

#[cfg(target_os = "macos")]
const DEFAULT_CONFIG_PATH: &str = "/Users/peter/webserve.ini";
#[cfg(target_os = "linux")]
const DEFAULT_CONFIG_PATH: &str = "/home/peter/webserve.ini";

impl Configuration {
    pub fn new() -> Configuration {
        Configuration{ worker_threads: 8,
                        enable_http: true,
                        port_number_http: 9393,
                        downgrade_user_after_bind: false,
                        downgrade_username: "".to_string(),
                        keep_alive_enabled: false,
                        keep_alive_timeout: 6,
                        keep_alive_limit: 20,
                        site_configs: Vec::with_capacity(0) }
    }

    pub fn auto_load_file(&mut self) -> bool {
        self.load_from_file(&DEFAULT_CONFIG_PATH)
    }

    // TODO: return Result instead...
    pub fn load_from_file(&mut self, filename: &str) -> bool {
        let file = std::fs::File::open(filename).unwrap();
        let reader = BufReader::new(file);

        // TODO: error handling...

        for line in reader.lines() {
            let line = line.unwrap();

            // ignore empty lines and comments
            if line.is_empty() || line.starts_with('#') {
                continue;
            }

            // first of all, see if it's a site definition statement
            if line.contains(" =") && line.contains("site(") {
                // it's a site config definition statement, i.e.:
                // <name> = site(<type>, <config>)
                // "photos = site(photos, host:photos.mydomain.net)

                self.site_configs.push(SiteConfig::default());
                let mut current_site_config = self.site_configs.last_mut().unwrap();

                let name_pos = line.find(' ').unwrap();
                let name = line[0..name_pos].to_string();

                current_site_config.name = name.clone();

                // println!("New Site: {}", name);

                if let Some(site_def_start) = line.find("site(") {
                    if let Some(site_def_end) = line.find(')') {
                        if site_def_start < site_def_end {
                            let site_def_start = site_def_start + 5;
                            let site_def_contents = line[site_def_start..site_def_end].to_string();

                            let (def_type, def_config) = site_def_contents.split_once(',').unwrap();
                            let def_type = def_type.trim();
                            let def_config = def_config.trim();
                            current_site_config.ctype = def_type.to_string();
                            current_site_config.definition = def_config.to_string();
                        }
                        else {
                            eprintln!("Warning: malformed site definition in config file...");
                        }
                    }
                    else {
                        eprintln!("Warning: malformed site definition in config file...");
                    }
                }

                // skip this line, we're done...
                continue;
            }

            // see if we're setting a param for a site, with a '.' in the initial name, i.e.
		    // photos.webContentPath: 1
            if line.contains(": ") {
                if let Some(colon_pos) = line.find(':') {
                    if let Some(dot_pos) = line.find('.') {
                        if dot_pos < colon_pos {
                            let full_name = line[0..colon_pos].to_string();

                            // site it belongs to - for the moment, we'll ignore this and use the last, but
                            // will need to fix this up in the future.
                            let _site_name = full_name[0..dot_pos].to_string();

                            let param_name = full_name[dot_pos + 1..].to_string();

                            let string_value = line[colon_pos + 1..].to_string();
                            let string_value = string_value.trim().to_string();

                            let current_site_config = self.site_configs.last_mut().unwrap();
                            current_site_config.params.insert(param_name, string_value);
                        }
                        else {
                            eprintln!("Warning: malformed site param definition in config file...");
                        }
                    }
                }
            }

            let mut key = String::new();
            let mut value = String::new();

            if !Configuration::get_key_value(&line, &mut key, &mut value) {
                // TODO: error
                continue;
            }

            if key == "workerThreads" {
                let int_value = value.parse::<u32>().unwrap();
                self.worker_threads = int_value;
            }
            else if Configuration::try_extract_bool_value("enableHTTP", &key, &value, &mut self.enable_http) {

            }
            else if key == "portNumberHTTP" {
                let int_value = value.parse::<u32>().unwrap();
                self.port_number_http = int_value;
            }
            else if Configuration::try_extract_bool_value("downgradeUserAfterBind", &key, &value, &mut self.downgrade_user_after_bind) {

            }
            else if key == "downgradeUserName" {
                if !value.is_empty() {
                    self.downgrade_username = value;
                }
            }
            else if Configuration::try_extract_bool_value("keepAliveEnabled", &key, &value, &mut self.keep_alive_enabled) {

            }
            else if key == "keepAliveTimeout" {
                let int_value = value.parse::<u32>().unwrap();
                self.keep_alive_timeout = int_value;
            }
            else if key == "keepAliveLimit" {
                let int_value = value.parse::<u32>().unwrap();
                self.keep_alive_limit = int_value;
            }
        }

        true
    }

    fn get_key_value(string_val: &str, key: &mut String, value: &mut String) -> bool {
        if !string_val.contains(':') {
            return false;
        }

        // this includes the separator, so...
//        let (key_str, val_str) = string_val.split_at(string_val.find(':').unwrap());
        let mut split = string_val.split(':');
        let (key_str, val_str) = (split.next().unwrap(), split.next().unwrap());

        // TODO: can we just do something like this?
        // key = *key_str.to_owned().to_string();

        key.clear();
        value.clear();

        key.push_str(key_str.trim());
        value.push_str(val_str.trim());

        true
    }

    fn get_bool_value_from_string(string_val: &str, bool_val: &mut bool) -> bool {
        if string_val.is_empty() {
            return false;
        }

        // check known strings first
        if string_val == "true" || string_val == "yes" {
            *bool_val = true;
            return true;
        }
        else if string_val == "false" || string_val == "no" {
            *bool_val = false;
            return true;
        }

        // otherwise, attempt to see if it's a number value, 0 or 1...
        let int_value = string_val.parse::<u32>().unwrap();
        if int_value == 0 {
            *bool_val = false;
            return true;
        }
        else if int_value == 1 {
            *bool_val = true;
            return true;
        }

        false
    }

    fn try_extract_bool_value(try_key_name: &str, actual_key_name: &str,
                              key_value: &str, config_field: &mut bool) -> bool {
        if try_key_name != actual_key_name {
            return false;
        }

        let mut extracted_bool_value = false;
        if Configuration::get_bool_value_from_string(key_value, &mut extracted_bool_value) {
            *config_field = extracted_bool_value;
        }
        else {
            // Error
        }

        // return that the key name actually matched, even if we didn't actually successfully apply the config value...
        true
    }
}

File: cpp/src/server/configuration.cpp
/*
 WebServe
 Copyright 2018-2022 Peter Pearson.

 Licensed under the Apache License, Version 2.0 (the "License");
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ---------
*/

#include "configuration.h"

#include "utils/string_helpers.h"

#include <fstream>

Configuration::Configuration() :
	m_workerThreads(16),
	m_enableHTTPv4(true),
	m_portNumberHTTPv4(9393),
	m_enableHTTPSv4(false),
	m_portNumberHTTPSv4(9394),
	m_redirectToHTTPS(false),
	m_enableHTTPv6(false),
	m_portNumberHTTPv6(9393),
	m_enableHTTPSv6(false),
	m_portNumberHTTPSv6(9394),
	m_enableHSTS(false),
	m_logOutputEnabled(true),
	m_logOutputTarget("stderr"),
	m_logOutputLevel("warning"),
	m_downgradeUserAfterBind(false),
	m_accessControlEnabled(false),
	m_closeBanClientsEnabled(false),
	m_closeBanClientsFailsThreshold(5),
	m_closeBanClientsTime(20),
	m_404NotFoundResponsesEnabled(true),
	m_keepAliveEnabled(true),
	m_keepAliveTimeout(6),
	m_keepAliveLimit(20),
	m_chunkedTransferJPEGsEnabled(false),
	m_sendDateHeaderField(true),
	m_tcpFastOpen(false)
{

}

bool Configuration::autoLoadFile()
{
#ifdef __linux__
	return loadFromFile("/home/peter/webserve.ini");
#else
	return loadFromFile("/Users/peter/webserve.ini");
#endif
}

bool Configuration::loadFromFile(const std::string& configPath)
{
	std::fstream fileStream(configPath.c_str(), std::ios::in);
	if (fileStream.fail())
	{
		return false;
	}

	std::string line;
	char buf[2048];

	std::string key;
	std::string value;
	
	SiteConfig* pCurrentSiteConfig = nullptr;

	while (fileStream.getline(buf, 2048))
	{
		line.assign(buf);

		// ignore empty lines and comments
		if (buf[0] == 0 || buf[0] == '#')
			continue;
		
		// first of all, see if it's a site definition statement
		if (line.find(" =") != std::string::npos && line.find("site(") != std::string::npos)
		{
			// it's a site config definition statement, i.e.:
			// <name> = site(<type>, <config>)
			// "photos = site(photos, host:photos.mydomain.net)
			
			m_aSiteConfigs.push_back(SiteConfig());
			
			pCurrentSiteConfig = &m_aSiteConfigs.back();
			
			size_t namePos = line.find(' ');
			std::string name = line.substr(0, namePos);
			
			pCurrentSiteConfig->m_name = name;
			
			size_t siteDefStart = line.find("site(");
			size_t siteDefEnd = line.find(')');
			
			if (siteDefStart != std::string::npos && siteDefEnd != std::string::npos &&
				siteDefStart < siteDefEnd)
			{
				siteDefStart += 5;
				
				std::string siteDefContents = line.substr(siteDefStart, siteDefEnd - siteDefStart);
				
				std::string siteDefType;
				std::string siteDefConfig;
				
				StringHelpers::splitInTwo(siteDefContents, siteDefType, siteDefConfig, ",");
				
				StringHelpers::stripWhitespace(siteDefType);
				StringHelpers::stripWhitespace(siteDefConfig);
				
				pCurrentSiteConfig->m_type = siteDefType;
				pCurrentSiteConfig->m_definition = siteDefConfig;
			}
			
			continue;
		}
		
		// see if we're setting a param for a site, with a '.' in the initial name, i.e.
		// photos.webContentPath: 1
		
		if (line.find(": ") != std::string::npos)
		{
			size_t colonPos = line.find(':');
			size_t dotPos = line.find('.');
			
			if (colonPos != std::string::npos && dotPos != std::string::npos &&
				dotPos < colonPos)
			{
				std::string fullName = line.substr(0, colonPos);
				
				// site it belongs to - for the moment, we'll ignore this and use the last, but
				// will need to fix this up in the future.
				std::string siteName = fullName.substr(0, dotPos);
				
				std::string paramName = fullName.substr(dotPos + 1);
				
				std::string stringValue = line.substr(colonPos + 1);
				StringHelpers::stripWhitespace(stringValue);
				
				pCurrentSiteConfig->m_aParams[paramName] = stringValue;
				
				continue;
			}
		}		

		if (!getKeyValue(line, key, value))
		{
			// TODO: error output
			continue;
		}

		// TODO: create some function infrastructure to handle this whole series of statements more concisely...
		if (key == "workerThreads")
		{
			unsigned int intValue = atoi(value.c_str());
			m_workerThreads = intValue;
		}
		else if (tryExtractBoolValue("enableHTTP", key, value, m_enableHTTPv4) || tryExtractBoolValue("enableHTTPv4", key, value, m_enableHTTPv4))
		{
			
		}
		else if (key == "portNumberHTTP" || key == "portNumberHTTPv4")
		{
			unsigned int intValue = atoi(value.c_str());
			m_portNumberHTTPv4 = intValue;
		}
		else if (tryExtractBoolValue("enableHTTPS", key, value, m_enableHTTPSv4) || tryExtractBoolValue("enableHTTPSv4", key, value, m_enableHTTPSv4))
		{
			
		}
		else if (key == "portNumberHTTPS" || key == "portNumberHTTPSv4")
		{
			unsigned int intValue = atoi(value.c_str());
			m_portNumberHTTPSv4 = intValue;
		}
		else if (tryExtractBoolValue("enableHTTPv6", key, value, m_enableHTTPv6))
		{

		}
		else if (key == "portNumberHTTPv6")
		{
			unsigned int intValue = atoi(value.c_str());
			m_portNumberHTTPv6 = intValue;
		}
		else if (tryExtractBoolValue("enableHTTPSv6", key, value, m_enableHTTPSv6))
		{

		}
		else if (key == "portNumberHTTPSv6")
		{
			unsigned int intValue = atoi(value.c_str());
			m_portNumberHTTPSv6 = intValue;
		}
		else if (key == "httpsCertificatePath")
		{
			// these https cert/key path items are a bit different, in that we allow multiple,
			// so we just emplace_back() them into the vector, rather than setting them directly.
			// TODO: the semantics within the .ini file for this are a bit odd, so maybe we
			//       use += syntax instead of the default?
			if (!value.empty())
			{
				m_httpsCertificatePaths.emplace_back(value);
			}
		}
		else if (key == "httpsKeyPath")
		{
			// these https cert/key path items are a bit different, in that we allow multiple,
			// so we just emplace_back() them into the vector, rather than setting them directly.
			// TODO: the semantics within the .ini file for this are a bit odd, so maybe we
			//       use += syntax instead of the default?
			if (!value.empty())
			{
				m_httpsKeyPaths.emplace_back(value);
			}
		}
		else if (tryExtractBoolValue("redirectToHTTPS", key, value, m_redirectToHTTPS))
		{
			
		}
		else if (tryExtractBoolValue("enableHSTS", key, value, m_enableHSTS))
		{
			
		}
		else if (tryExtractBoolValue("logOutputEnabled", key, value, m_logOutputEnabled))
		{

		}
		else if (key == "logOutputLevel")
		{
			if (!value.empty())
			{
				m_logOutputLevel = value;
			}
			else
			{
				// error, and/or turn off downgradeUserAfterBind ?
			}
		}
		else if (key == "logOutputTarget")
		{
			if (!value.empty())
			{
				m_logOutputTarget = value;
			}
		}
		else if (tryExtractBoolValue("downgradeUserAfterBind", key, value, m_downgradeUserAfterBind))
		{

		}
		else if (key == "downgradeUserName")
		{
			if (!value.empty())
			{
				m_downgradeUserName = value;
			}
			else
			{
				// error, and/or turn off downgradeUserAfterBind ?
			}
		}
		else if (tryExtractBoolValue("accessControlEnabled", key, value, m_accessControlEnabled))
		{

		}
		else if (key == "accessControlLogTarget")
		{
			if (!value.empty())
			{
				m_accessControlLogTarget = value;
			}
		}
		else if (tryExtractBoolValue("closeBanClientsEnabled", key, value, m_closeBanClientsEnabled))
		{

		}
		else if (key == "closeBanClientsFailsThreshold")
		{
			unsigned int intValue = atoi(value.c_str());
			m_closeBanClientsFailsThreshold = intValue;
		}
		else if (key == "closeBanClientsTime")
		{
			unsigned int intValue = atoi(value.c_str());
			m_closeBanClientsTime = intValue;
		}
		else if (tryExtractBoolValue("404NotFoundResponsesEnabled", key, value, m_404NotFoundResponsesEnabled))
		{

		}
		else if (tryExtractBoolValue("keepAliveEnabled", key, value, m_keepAliveEnabled))
		{

		}
		else if (key == "keepAliveTimeout")
		{
			unsigned int intValue = atoi(value.c_str());
			m_keepAliveTimeout = intValue;
		}
		else if (key == "keepAliveLimit")
		{
			unsigned int intValue = atoi(value.c_str());
			m_keepAliveLimit = intValue;
		}
		else if (tryExtractBoolValue("sendDateHeaderField", key, value, m_sendDateHeaderField))
		{

		}
		else if (tryExtractBoolValue("chunkedTransferJPEGsEnabled", key, value, m_chunkedTransferJPEGsEnabled))
		{

		}
		else if (tryExtractBoolValue("tcpFastOpen", key, value, m_tcpFastOpen))
		{

		}
		else
		{
			// TODO: proper logging, but we currently don't have a logger at this point, so...
			fprintf(stderr, "Warning: Unrecognised configuration item: '%s' in WebServe config file...\n", key.c_str());
		}
	}

	fileStream.close();

	return true;
}

bool Configuration::tryExtractBoolValue(const std::string& tryKeyName, const std::string& actualKeyName,
										const std::string& keyValue, bool& configMemberVariable)
{
	if (tryKeyName != actualKeyName)
		return false;

	bool extractedBoolValue = false;
	if (getBoolValueFromString(keyValue, extractedBoolValue))
	{
		configMemberVariable = extractedBoolValue;
	}
	else
	{
		// TODO: error...
	}

	// return that the key name actually matched, even if we didn't actually successfully apply the config value...
	return true;
}

bool Configuration::getKeyValue(const std::string& configLine, std::string& key, std::string& value)
{
	size_t sepPos = configLine.find(':');
	if (sepPos == std::string::npos)
		return false;

	size_t valueStart = configLine.find_first_not_of(' ', sepPos + 1);
	if (valueStart == std::string::npos)
		return false;

	key = configLine.substr(0, sepPos);
	value = configLine.substr(valueStart);

	return true;
}

// note: this returns whether it extracted a value or not, not the actual value!!!
bool Configuration::getBoolValueFromString(const std::string& stringValue, bool& boolValue)
{
	if (stringValue.empty())
		return false;

	// check known strings first
	if (stringValue == "true" || stringValue == "yes")
	{
		boolValue = true;
		return true;
	}
	else if (stringValue == "false" || stringValue == "no")
	{
		boolValue = false;
		return true;
	}

	// otherwise attempt to see if it's a number value, 0 or 1...
	// TODO: obviously atoi() might return 0 on an invalid string, so this isn't really robust...
	unsigned int intValue = atoi(stringValue.c_str());
	if (intValue == 0)
	{
		boolValue = false;
		// TODO: this isn't really robust, so somewhat makes a mockery of doing it this way...
		return true;
	}
	else if (intValue == 1)
	{
		boolValue = true;
		return true;
	}

	return false;
}
